File: leetcode0198/solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        prev_prev= 0
        prev = nums[0]

        for i in range(1, len(nums)):
            prev_prev, prev = prev, max(prev, prev_prev + nums[i])

        return prev

# class Solution:
#     def rob(self, nums: List[int]) -> int:
#         if(len(nums) <= 2):
#             return max(nums)
#         if len(nums) % 2:
#             nums.append(0)
#         for i in range(1, int(len(nums) / 2)):
#             if i >= 2:
#                 nums[i * 2] += max(nums[(i - 1) * 2], nums[(i - 2) * 2 + 1])
#             else:
#                 nums[i * 2] += nums[(i - 1) * 2]
#             nums[i * 2 + 1] += max(nums[(i - 1) * 2] , nums[(i - 1) * 2 + 1])

          
#         return max(nums[-1], nums[-2])

File: leetcode0198/solution.rs
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let mut n: usize = nums.len();

        let (mut prev_prev, mut prev) = (0_i32, 0_i32);

        for i in 0..n {
            (prev_prev, prev) = (prev, prev.max(prev_prev + nums[i]));
        }

        prev
    }
}

// impl Solution {
//     pub fn rob(nums: Vec<i32>) -> i32 {
//         let mut n: usize = nums.len();

//         let mut dp: Vec<i32> = vec![0; n + 2];

//         for i in 0..n {
//             dp[i + 2] = dp[i + 1].max(dp[i] + nums[i]);
//         }

//         *dp.last().unwrap()
//     }
// }